src/components/inspectors/TimerExpression.vue

```diff
@@ -0,0 +1,381 @@
+        <option value="day">day</option>
+        <option value="week">week</option>
+        <option value="month">month</option>
+        <option value="year">year</option>
+      </select>
+    </div>
+    </template>
+    <div v-if="periodicity==='week'">
+      <label>{{weekLabel}}</label>
+      <div>
+        <span v-for="(day, index) in weekdays" :key="index + 'week'" 
+              class="badge badge-pill weekday"
+              :class="{'badge-primary': day.selected && !sameDay, 'badge-light': !(day.selected && !sameDay), 'border border-primary': currentDaySelected(day)}"
+              @click="clickWeekDay(day);update()">{{day.initial}}</span>
+      </div>
+    </div>
+    <template v-if="hasEnds">
+      <label>Ends</label>
+      <div>
+        <div class="form-check">
+          <label class="form-check-label">
+            <input type="radio" class="form-check-input" name="optradio" value="never" v-model="ends" @change="update">Never
+          </label>
+        </div>
+        <div class="form-check check-input">
+          <label class="form-check-label">
+            <input type="radio" class="form-check-input" name="optradio" value="ondate" v-model="ends" @change="update">On &nbsp;
+          </label>
+          <datepicker v-model="endDate" calendar-class="calendar calendaron" :disabled="ends!=='ondate'" format="yyyy-MM-dd"
+                      input-class="form-control end-date"
+                      class=" control float-right"
+                      :class="{'date-disabled' : ends!=='ondate'}"
+                      @selected="updateEndDate"></datepicker>
+        </div>
+        <div class="form-check check-input">
+          <label class="form-check-label">
+            <input type="radio" class="form-check-input" name="optradio" value="after" v-model="ends" @change="update">After &nbsp;
+          </label>
+          <input v-model="times" type="number" min="0" :disabled="ends!=='after'" 
+                 class="form-control control after float-right" @change="update">
+          <label class="occurrences">occurrences</label>
+        </div>
+      </div>
+    </template>
+  </div>
+</template>
+
+<script>
+import Datepicker from 'vuejs-datepicker';
+import moment from 'moment';
+
+const periods = {
+  'day': 'D',
+  'week': 'W',
+  'month': 'M',
+  'year': 'Y',
+};
+
+export default {
+  components: {
+    Datepicker,
+  },
+  props: {
+    value: String,
+    hasEnds: {
+      type: Boolean,
+      default: true,
+    },
+    hasRepeat: {
+      type: Boolean,
+      default: true,
+    },
+    repeatLabel: {
+      type: String,
+      default: 'Repeat every',
+    },
+    weekLabel: {
+      type: String,
+      default: 'Repeat on',
+    },
+  },
+  data() {
+    const date = moment().set('hour', 0).set('minutes', 0);
+    const hours = [];
+    for (let i = 0; i < 48; i++) {
+      hours.push(date.format('HH:mm'));
+      date.add(30, 'minutes');
+    }
```

Comment:
The code to get the list of hours above isn't very readable. I think you can do two things to clean it up:
1. Use a hardcoded list of hours if this is something that is unlikely to change or be configurable.
2. Extract it into a function above and call that. Something like:
```js
function getListOfHalfHourIntervalsInADay() {
  const date = moment().startOf('day');
  const dateEnd = moment().endOf('day');
  const intervals = [];

  while (date < dateEnd ) {
    intervals.push(date.format('HH:mm'));
    date.add(30, 'minutes');
  }

  return intervals;
}

export default {
  ...
  data() {
    hours: getListOfHalfHourIntervalsInADay(),
    ...
  }
  ...
```